database/pgsql/ancestry.go

```diff
@@ -43,80 +43,89 @@ func (tx *pgSession) UpsertAncestry(ancestry database.Ancestry, features []datab
-	}
+	layers := []database.Layer{}
+	for rows.Next() {
+		var (
+			hash    string
+			layer   database.Layer
+			isValid bool
+			err     error
+		)
 
-	processed.Detectors, err = tx.findProcessors(searchAncestryDetectors, "searchAncestryDetectors", "detector", ancestryID)
-	if err != nil {
-		return ancestry, processed, false, err
-	}
+		if err = rows.Scan(&hash); err != nil {
+			return ancestry, false, handleError("searchAncestryLayer", err)
+		}
 
-	processed.Listers, err = tx.findProcessors(searchAncestryListers, "searchAncestryListers", "lister", ancestryID)
-	if err != nil {
-		return ancestry, processed, false, err
+		if layer, isValid, err = tx.FindLayer(hash); err != nil || !isValid {
+			return ancestry, isValid, err
+		}
+
+		layers = append(layers, layer)
 	}
 
-	return ancestry, processed, true, nil
+	ancestry.Layers = layers
+	return ancestry, true, nil
 }
 
-func (tx *pgSession) FindAncestryFeatures(name string) (database.AncestryWithFeatures, bool, error) {
+func (tx *pgSession) FindAncestryWithContent(name string) (database.AncestryWithContent, bool, error) {
 	var (
-		awf database.AncestryWithFeatures
-		ok  bool
-		err error
+		ancestryContent  database.AncestryWithContent
+		ancestryMetadata database.Ancestry
+		isValid          bool
+		err              error
 	)
-	awf.Ancestry, awf.ProcessedBy, ok, err = tx.FindAncestry(name)
-	if err != nil {
-		return awf, false, err
-	}
 
-	if !ok {
-		return awf, false, nil
+	if ancestryMetadata, isValid, err = tx.FindAncestry(name); err != nil || !isValid {
+		return ancestryContent, isValid, err
 	}
 
 	rows, err := tx.Query(searchAncestryFeatures, name)
 	if err != nil {
-		return awf, false, handleError("searchAncestryFeatures", err)
+		return ancestryContent, false, handleError("searchAncestryFeatures", err)
 	}
 
+	features := map[int][]database.NamespacedFeature{}
 	for rows.Next() {
-		nf := database.NamespacedFeature{}
-		err := rows.Scan(&nf.Namespace.Name, &nf.Namespace.VersionFormat, &nf.Feature.Name, &nf.Feature.Version)
-		if err != nil {
-			return awf, false, handleError("searchAncestryFeatures", err)
+		var (
+			feature database.NamespacedFeature
+			// layerIndex is used to determine which layer the namespaced feature belongs to.
+			layerIndex sql.NullInt64
+		)
+
+		if err := rows.Scan(&feature.Namespace.Name,
+			&feature.Namespace.VersionFormat,
+			&feature.Feature.Name, &feature.Feature.Version,
+			&layerIndex); err != nil {
+			return ancestryContent, false, handleError("searchAncestryFeatures", err)
 		}
-		nf.Feature.VersionFormat = nf.Namespace.VersionFormat
-		awf.Features = append(awf.Features, nf)
+
+		feature.Feature.VersionFormat = feature.Namespace.VersionFormat // This looks strange.
+		features[int(layerIndex.Int64)] = append(features[int(layerIndex.Int64)], feature)
 	}
 
-	return awf, true, nil
+	// By the assumption of Ancestry Layer Index, we have the ancestry's layer
+	// index corresponding to the index in the array.
```

Comment:
This should really be in the Session interface doc comment for FindAncestryWithContent.

Reply:
Well, this is an implementation detail for the ancestry_layer table. Here, I'm using the `ancestry_index` as the key to associate features to ancestry layers.

For people who implements this interface, they don't need to know what `ancestry_index` is. All they should do is get corresponding features for each layer in the Ancestry. They can do it in different ways, for example, query layer-wise feature directly using a single query.

Reply:
You're totally right.